### DISCUSSION TITLE
Windscribe OpenVPN TLS Verification Errors

### DISCUSSION THREAD

**Discussion by granroth** [qdm12/gluetun#619] [Q&A] [answered]
I've been using `gluetun` as part of a `docker-compose` setup on a Raspberry Pi 4 successfully for a bit over a year, now. At some point in the past few weeks, though, it has stopped working for me and instead throws TLS verification errors. In this time, I did `docker-compose pull` to get the latest version of all images, but had not changed the configuration files at all.

The relevant errors look like so:
```
gluetun      | 2021/09/12 16:53:31 INFO openvpn: TLS: Initial packet from [AF_INET]104.129.3.67:1194, sid=bba06001 09f0c526
gluetun      | 2021/09/12 16:53:31 INFO openvpn: VERIFY ERROR: depth=2, error=format error in certificate's notBefore field: C=CA, ST=ON, L=Toronto, O=Windscribe Limited, OU=Systems, CN=Windscribe Node CA X1, serial=529079670295837910636462923234892496861000910496
gluetun      | 2021/09/12 16:53:31 INFO openvpn: OpenSSL: error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
gluetun      | 2021/09/12 16:53:31 INFO openvpn: TLS_ERROR: BIO read tls_read_plaintext error
gluetun      | 2021/09/12 16:53:31 INFO openvpn: TLS Error: TLS object -> incoming plaintext read error
gluetun      | 2021/09/12 16:53:31 INFO openvpn: TLS Error: TLS handshake failed
gluetun      | 2021/09/12 16:53:31 INFO openvpn: TCP/UDP: Closing socket
```
I do have a Windscribe Pro account and so submitted a help ticket on their end. As one would expect, their support procedures only cover using `windscribe-cli` and `openvpn` with the generated configs, and certainly do not cover `gluetun`. The one snippet of info dropped by the service tech was that Windscribe no longer supported the `compress` flag, since a few months ago, and that might be the root cause. I see no reference to anything related to a `compress` flag in `gluetun`.

I'm a bit at a loss, if only because I'm not seeing any other references to this error, yet if it was a server-end change like the Windscribe blog suggests, then I would expect that all `gluetun` Windscribe users would be seeing this... but they are not. That would indicate a config error on my part, yet I haven't changed my config in over a year and it was previously working perfectly.

Any idea where I can look next?

**Comment by qdm12** [comment]
The Windscribe OpenVPN configuration file is built [here](https://github.com/qdm12/gluetun/blob/40342619e727e81d017528271baa8162d438e566/internal/provider/windscribe/openvpnconf.go#L14-L80) in the code. Under the hood, gluetun just uses an official build of OpenVPN in the end (contrary to Wireguard which is custom-ish built). You can see the resulting OpenVPN configuration file used with:

```sh
docker run -it --rm -e VPNSP=windscribe -e OPENVPN_USER=bla -e OPENVPN_PASSWORD=bla qmcgaw/gluetun openvpnconfig
```

Which is for example:

```
client
dev tun0
nobind
persist-key
remote-cert-tls server
ping 10
ping-exit 60
ping-timer-rem
tls-exit
key-direction 1
script-security 2
reneg-sec 0
ncp-disable
auth-nocache
mute-replay-warnings
pull-filter ignore "auth-token"
auth-retry nointeract
suppress-timestamps
verb 1
auth-user-pass /etc/openvpn/auth.conf
proto udp
remote 185.104.187.43 1194
auth sha512
verify-x509-name bud-90.windscribe.com name
data-ciphers-fallback aes-256-cbc
data-ciphers aes-256-cbc
pull-filter ignore "route-ipv6"
pull-filter ignore "ifconfig-ipv6"
<ca>
-----BEGIN CERTIFICATE-----
MIIF5zCCA8+gAwIBAgIUXKzAwOtQBNDoTXcnwR7GxbVkRqAwDQYJKoZIhvcNAQELBQAwezELMAkGA1UEBhMCQ0ExCzAJBgNVBAgMAk9OMRAwDgYDVQQHDAdUb3JvbnRvMRswGQYDVQQKDBJXaW5kc2NyaWJlIExpbWl0ZWQxEDAOBgNVBAsMB1N5c3RlbXMxHjAcBgNVBAMMFVdpbmRzY3JpYmUgTm9kZSBDQSBYMTAeFw0yMTA3MDYyMTM5NDNaFw0zNzA3MDIyMTM5NDNaMHsxCzAJBgNVBAYTAkNBMQswCQYDVQQIDAJPTjEQMA4GA1UEBwwHVG9yb250bzEbMBkGA1UECgwSV2luZHNjcmliZSBMaW1pdGVkMRAwDgYDVQQLDAdTeXN0ZW1zMR4wHAYDVQQDDBVXaW5kc2NyaWJlIE5vZGUgQ0EgWDEwggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQDg/79XeOvthNbhtocxaJ6raIsrlSrnUJ9xAyYHJV+auT4ZlACNE54NVhrGPBEVdNttUdezHaPUlQA+XTWUPlHMayIg9dsQEFdHH3StnFrjYBzeCO76trPZ8McU6PzW+LqNEvFAwtdKjYMgHISkt0YPUPdB7vED6yqbyiIAlmN5u/uLG441ImnEq5kjIQxVB+IHhkV4O7EuiKOEXvsKdFzdRACi4rFOq9Z6zK2Yscdg89JvFOwIm1nY5PMYpZgUKkvdYMcvZQ8aFDaArniu+kUZiVyUtcKRaCUCyyMM7iiN+5YV0vQ0Etv59ldOYPqL9aJ6QeRG9Plq5rP8ltbmXJRBO/kdjQTBrP4gYddt5W0uv5rcMclZ9te0/JGl3Os3Gps5w7bYHeVdYb3j0PfsJAQ5WrM+hS5/GaX3ltiJKXOA9kwtDG3YpPqvpMVAqpM6PFdRwTH62lOemVAOHRrThOVbclqpEbe3zH59jwSML5WXgVIfwrpcpndj2uEyKS50y30GzVBIn5M1pcQJJplYuBp8nVGCqA9AVV+JHffVP/JrkvEJzhui8M5SVnkzmAK3i+rwL0NMRJKwKaSm1uJVvJyoXMMNTEcu1lqnSl+i2UlIYAgeqeT//D9zcNgcOdP8ix6NhFChjE1dvNFv8mXxkezmu+etPpQZTpgc1eBZvAAojwIDAQABo2MwYTAdBgNVHQ4EFgQUVLNKLT/c9fTG4BJ+6rTZkPjS4RgwHwYDVR0jBBgwFoAUVLNKLT/c9fTG4BJ+6rTZkPjS4RgwDwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMCAYYwDQYJKoZIhvcNAQELBQADggIBAF4Bpc0XdBsgF3WSeRLJ6t2J7vOjjMXBePwSL0g6GDjLpKW9sz9F3wfXaK5cKjY5tj5NEwmkVbqa+BXg4FWic0uLinI7tx7sLtvqHrKUFke35L8gjgIEpErg8nmBPokEVsmCcfYYutwOi2IGikurpY29O4HniDY9baXp8kvwn1T92ZwF9G5SGzxc9Y0rGs+BwmDZu58IhID3aqAJ16aHw5FHQWGUxje5uNbEUFdVaj7ODvznM6ef/5sAFVL15mftsRokLhCnDdEjI/9QOYQoPrKJAudZzbWeOux3k93SehS7UWDZW4AFz/7XTaWL79tLqqtTI6LiuHn73enHgH6BlsH3ESB+Has6Rn7aH0wBByLQ9+NYIfAwXUCd4nevUXeJ3r/aORi367ATj1yb3J8llFCsoc/PT7a+PxDT8co2m6TtcRK3mFT/71svWB0zy7qAtSWT1C82W5JFkhkP44UMLwGUuJsrYy2qAZVru6Jp6vU/zOghLp5kwa1cO1GEbYinvoyTw4XkOuaIfEMUZA10QCCW8uocxqIZXTzvF7LaqqsTCcAMcviKGXS5lvxLtqTEDO5rYbf8n71J2qUyUQ5yYTE0UFQYiYTuvCbtRg2TJdQy05nisw1O8Hm2erAmUveSTr3CWZ/av7Dtup352gRS6qxW4w0jRN3NLfLyazK/JjTX
-----END CERTIFICATE-----
</ca>
<tls-auth>
-----BEGIN OpenVPN Static key V1-----
5801926a57ac2ce27e3dfd1dd6ef82042d82bd4f3f0021296f57734f6f1ea714a6623845541c4b0c3dea0a050fe6746cb66dfab14cda27e5ae09d7c155aa554f399fa4a863f0e8c1af787e5c602a801d3a2ec41e395a978d56729457fe6102d7d9e9119aa83643210b33c678f9d4109e3154ac9c759e490cb309b319cf708cae83ddadc3060a7a26564d1a24411cd552fe6620ea16b755697a4fc5e6e9d0cfc0c5c4a1874685429046a424c026db672e4c2c492898052ba59128d46200b40f880027a8b6610a4d559bdc9346d33a0a6b08e75c7fd43192b162bfd0aef0c716b31584827693f676f9a5047123466f0654eade34972586b31c6ce7e395f4b478cb
-----END OpenVPN Static key V1-----
</tls-auth>
```

Do you see an obvious difference with one of their official openvpn config regarding:

1. `verify-x509-name`?
2. The long CA value `<ca>`?
3. The long TLS Auth value `<tls-auth>`?

If not, please comment back with one of their Openvpn config so I can see if anything changed. It's also normal there are differences with the official config since the one used by Gluetun is a bit customized (like `suppress-timestamps` or `pull-filter ignore "route-ipv6"`) on top.

**Comment by granroth** [comment]
Here's my `gluetun` generated ovpn file vs the one generated by Windscribe for a local Phoenix (closest geographically) server - re-ordered only to make `diff` easier:

![image](https://user-images.githubusercontent.com/1722048/133012942-8f5c9f1b-0a04-46d1-bfda-b3e153d80467.png)

There _are_ differences but honestly not as many as I'd assume. The `<ca>` and `<tls-auth>` sections are identical. There are a number of "extra" `gluetun` configs. The Windscribe config has two extra `cipher`-related entries.

The `remote` and `verify-x509-name` settings both differ -- The Windscribe config corresponds to `us-west-046.whiskergalaxy.com` while the `gluetun` config is giving me a Tel Aviv, Isreal set. Maybe worth noting that when I run `gluetun` for real, I have a `SERVER_HOSTNAME` set to a bunch of US-based servers. I did attempt to restrict it only to `- SERVER_HOSTNAME=us-west-046.whiskergalaxy.com` to match the official config, but that made no difference.

**Reply by qdm12** [reply]
Oh silly me! I didn't pay attention to the `Raspberry Pi` part! I guess you are running Raspbian right? If so, you are running a 32 bit OS, so check out https://github.com/qdm12/gluetun/wiki/32-bit-setup! I'm 95% sure this will solve it!

If it does solve it, can you please report back if running with `-e OPENVPN_CIPHER=aes-256-gcm` work? (since that's what they specific in their config). You could also just run a native openvpn client with that config file you have on the left. Thanks!

In the meantime, I'll clean up some options from the generated configuration file like `reneg-sec 0` or `script-security 2` which are useless.

**Reply by granroth** [reply]
That did it! I wouldn't have thought to look at the "32-bit-setup" page since I had already set it up a year ago successfully and I wouldn't have thought to upgrade `libseccomp2` since I already had to do that to get Radarr to run... but, indeed, upgrading `libseccomp2` outside of the backports repo and directly to 2.5.1-1 solved my problem like magic.

Thank you!

(I will mark your first reply as the 'answer' since your recent reply with the working answer doesn't have such an option)

**Reply by qdm12** [reply]
Nice! Thanks, sorry to repeat, but now that it works I need some tester for Windscribe 😄

> If it does solve it, can you please report back if running with -e OPENVPN_CIPHER=aes-256-gcm work? (since that's what they specific in their config). You could also just run a native openvpn client with that config file you have on the left. Thanks!

Also side note you could use wireguard too which is well supported for Windscribe.

**Reply by granroth** [reply]
Okay, I tried my now-working setup with `OPENVPN_CIPHER=aes-256-gcm` and it worked like a charm. With no special cipher, my logs showed that I was using AES-CBC + HMAC:
```
INFO openvpn: Outgoing Data Channel: Cipher 'AES-256-CBC' initialized with 256 bit key
INFO openvpn: Outgoing Data Channel: Using 512 bit message hash 'SHA512' for HMAC authentication
```
With the specific GCM cipher option, everything continues to work, but now I see:
```
INFO openvpn: Outgoing Data Channel: Cipher 'AES-256-GCM' initialized with 256 bit key
```
Both result in a working VPN.

I will test Wireguard in a separate reply.

**Reply by granroth** [reply]
I had initially shied away from Wireguard since I prefer to start from a solid working base before branching out into new software or methods. Now that I do have a working OpenVPN-based VPN again, though, it's safe to experiment!

I started by generating a config file from the Windscribe site. The end result looks like so:
```
[Interface]
PrivateKey = <PRIVATE_KEY>
Address = <IP/MASK>
DNS = <IP>

[Peer]
PublicKey = <PUBLIC_KEY>
AllowedIPs = 0.0.0.0/0
Endpoint = phx-325-wg.whiskergalaxy.com:65142
PresharedKey = <PRESHARED_KEY>
```
I blocked out most of the values because, to be honest, I'm not 100% certain which are truly shared and which are specific to my account. So the (unused) `Endpoint` is the only value I'm comfortable sharing.

These fields map to `gluetun` like so:
```
[Interface]PrivateKey = WIREGUARD_PRIVATE_KEY
[Interface]Address = WIREGUARD_ADDRESS
[Peer]PresharedKey = WIREGUARD_PRESHARED_KEY
```
Those are the only three that were absolutely necessary. Thus, my `gluetun` config looks like:
```
      - VPNSP=windscribe
      - TZ=America/Phoenix
      - VPN_TYPE=wireguard
      - WIREGUARD_PRIVATE_KEY=<PRIVATE_KEY>
      - WIREGUARD_ADDRESS=<ADDRESS>/32
      - WIREGUARD_PRESHARED_KEY=<PRESHARED_KEY>
```
And _voila_, it worked like a charm!

I did not set any explicit servers and it seems to just pick a random one, worldwide.

Worth saying that I wanted to run some speed tests comparing OpenVPN and Wireguard, but I'm currently drawing a blank at an effective way of doing so. I tried a simple:
```
time docker run --rm --network=container:gluetun alpine:3.13 wget -O /dev/null -q http://speedtest.newark.linode.com/100MB-newark.bin
```
That was ineffective, since the resulting times were all over the map. An expected time of sub-10s ranged from 40s to well over a minute, for either protocol. Those look like unreliable results.

**Reply by qdm12** [reply]
@granroth Perfect, thank you so much for trying all this and taking the time again to report all this!

- [x] I'll change the default cipher to `aes-256-gcm` as per their openvpn configurations then. - done in a1588302a7fbd9b22470f0e2235507c36c139021


> I had initially shied away from Wireguard since I prefer to start from a solid working base before branching out into new software

I totally understand, although - to be fair - OpenVPN has been quite more unstable than Wireguard so far in my experience 😄 Wireguard is just a lot more dumb especially in terms of options (no need for CAs, openssl and all that).

> I started by generating a config file from the Windscribe site

As far as I know, you can use those private key + address + pre-shared key for all their servers, so it's really set-once environment variables 😉 

> I blocked out most of the values

Private key and pre-shared key are the ones to protect. The rest is useless (assigned address) or public information (endpoint, public key). All the Wireguard public keys of each server are actually bundled in the gluetun program ([here](https://github.com/qdm12/gluetun/blob/91ce790b6b7412b7c83d14133617c613d6d07d51/internal/storage/servers.json)), that's why you don't need to set them 👍 (thumbing up myself I guess?)

> I did not set any explicit servers and it seems to just pick a random one, worldwide.

Yes you can filter by `REGION`, and/or `CITY` and/or `SERVER_HOSTNAME`. Gluetun will pick one of the filtered out servers at (pseudo-)random.

> but I'm currently drawing a blank at an effective way of doing so

Oh boy, I need to do this! I spent months implementing Wireguard bit by bit, and I haven't even ran a speedtest, I'm so ashamed. Anyway I did it!

Command I used:

```sh
docker run -it --rm --network=container:gluetun alpine:3.14 /bin/sh -c "apk add speedtest-cli && speedtest-cli"
```

Gluetun is connected to the nearest VPN server.

Two machines/config tested:

1. Docker Desktop on WSL2 on a Windows host
    - 24 threads cpu (amd 5900x)
    - Running on Docker desktop with wsl2 on a windows host
2. Linux host
    - 12 threads cpu (amd 2600x)
    - Running on an Arch Linux host with recent kernel

| Machine | Protocol | Download | Upload | Download diff vs other protocol |
| --- | --- | --- | --- | --- |
| Windows host | OpenVPN | `299.24` | `463.69` | +0.23% |
| Windows host | Wireguard | `298.55` | `493.49` | -0.23% |
| Linux host | OpenVPN | `645.91` | `481.25` | -4% |
| Linux host | Wireguard | `673.37` | `561.28` | +4.2% |

Conclusion: Docker desktop on Windows sucks network performance wise. Wireguard is a tiny bit faster on both hosts. Definitely not worth the effort 😄 Although maybe on a low power device with a lot of cores and low single thread performance, it might make a more significant difference.

Added that performance test to the Wiki [here](https://github.com/qdm12/gluetun/wiki/Wireguard#performance)

**Reply by granroth** [reply]
Both OpenVPN and Wireguard in Docker take their tool on the CPU and, as a result, are _significantly_ slower than no VPN at all on my Raspberry Pi 4. Wireguard is showing faster speeds in isolated tests, but not by notable amounts. In small tests, I've seen OpenVPN be stable over time while Wireguard wants to drop the connection (at least DNS resolution) without restarting after only a few hours. I'm going to test Wireguard a bit more to see if that was a fluke.

Protocol | Download | Upload
-------- | -------- | ------
None | 176 | 10
Wireguard | 57 | 10
OpenVPN | 50 | 10

**Reply by qdm12** [reply]
Thanks I added your results to that wiki page 🚀

**Comment by qdm12** [comment, marked as answer]
Answer [**in this sub-reply**](https://github.com/qdm12/gluetun/discussions/619#discussioncomment-1316960)
